src/borg/testsuite/repo_new.py

```diff
@@ -369,3 +396,488 @@ def test_crash_before_write_index(repository, reopened_repository):
+
+
+""" QUOTA TESTS """
+
+
+def test_tracking(repository, reopened_repository):
+    assert repository.storage_quota_use == 0
+    ch1 = fchunk(bytes(1234))
+    repository.put(H(1), ch1)
+    assert repository.storage_quota_use == len(ch1) + 41 + 8
+    ch2 = fchunk(bytes(5678))
+    repository.put(H(2), ch2)
+    assert repository.storage_quota_use == len(ch1) + len(ch2) + 2 * (41 + 8)
+    repository.delete(H(1))
+    assert repository.storage_quota_use == len(ch1) + len(ch2) + 2 * (41 + 8)  # we have not compacted yet
+    repository.commit(compact=False)
+    assert repository.storage_quota_use == len(ch1) + len(ch2) + 2 * (41 + 8)  # we have not compacted yet
+    with reopened_repository as repository:
+        # Open new transaction; hints and thus quota data is not loaded unless needed.
+        ch3 = fchunk(b"")
+        repository.put(H(3), ch3)
+        repository.delete(H(3))
+        assert repository.storage_quota_use == len(ch1) + len(ch2) + len(ch3) + 3 * (
+            41 + 8
+        )  # we have not compacted yet
+        repository.commit(compact=True)
+        assert repository.storage_quota_use == len(ch2) + 41 + 8
+
+
+def test_exceed_quota(repository, reopened_repository):
+    assert repository.storage_quota_use == 0
+    repository.storage_quota = 80
+    ch1 = fchunk(b"x" * 7)
+    repository.put(H(1), ch1)
+    assert repository.storage_quota_use == len(ch1) + 41 + 8
+    repository.commit(compact=False)
+    with pytest.raises(Repository.StorageQuotaExceeded):
+        ch2 = fchunk(b"y" * 13)
+        repository.put(H(2), ch2)
+    assert repository.storage_quota_use == len(ch1) + len(ch2) + (41 + 8) * 2  # check ch2!?
+    with pytest.raises(Repository.StorageQuotaExceeded):
+        repository.commit(compact=False)
+    assert repository.storage_quota_use == len(ch1) + len(ch2) + (41 + 8) * 2  # check ch2!?
+    with reopened_repository as repository:
+        repository.storage_quota = 150
+        # Open new transaction; hints and thus quota data is not loaded unless needed.
+        repository.put(H(1), ch1)
+        assert (
+            repository.storage_quota_use == len(ch1) * 2 + (41 + 8) * 2
+        )  # we have 2 puts for H(1) here and not yet compacted.
+        repository.commit(compact=True)
+        assert repository.storage_quota_use == len(ch1) + 41 + 8  # now we have compacted.
+
+
+""" AUXILIARY CORRUPTION TESTS """
+
+
+@pytest.fixture()
+def auxiliary_repository(repository):
+    repository.put(H(0), fchunk(b"foo"))
+    repository.commit(compact=False)
+    repository.close()
+    yield repository
+
+
+def do_commit(repository):
+    with repository:
+        repository.put(H(0), fchunk(b"fox"))
+        repository.commit(compact=False)
+
+
+def test_corrupted_hints(auxiliary_repository):
+    with auxiliary_repository as repository:
+        with open(os.path.join(repository.path, "hints.1"), "ab") as fd:
+            fd.write(b"123456789")
+        do_commit(repository)
```

Comment:
This test gives a warning, posted below. If I comment out this test, then the warning going on to the next test, and so on. I believe this has to do with an issue with my auxiliary_repository.

```
src/borg/testsuite/repo_new.py::test_corrupted_hints
  /~/python3.11/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: <function Repository.__del__ at 0x102140d60>
  
  Traceback (most recent call last):
    File "/Users/ted/Desktop/borg_workspace/borg/src/borg/repository.py", line 211, in __del__
      self.close()
    File "/Users/ted/Desktop/borg_workspace/borg/src/borg/repository.py", line 488, in close
      self.lock.release()
    File "/Users/ted/Desktop/borg_workspace/borg/src/borg/locking.py", line 431, in release
      self._roster.modify(EXCLUSIVE, REMOVE)
    File "/Users/ted/Desktop/borg_workspace/borg/src/borg/locking.py", line 329, in modify
      elements.remove(self.id)
  KeyError: ('theodores-macbook-air.local@169793140525769', 43199, 0)
  
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
```